doc/epkg.1
.TH epkg 1 "Oct 2002" "University of Illinois" "Package Management"
.SH NAME
epkg \- Encap package manager
.SH SYNOPSIS
\fBepkg [options] [-i] pkgspec ...\fP
.br
\fBepkg [options] -r pkgspec ...\fP
.br
\fBepkg [options] -k pkgspec ...\fP
.br
\fBepkg [options] -u [pkgspec ...]\fP
.br
\fBepkg [options] -b\fP
.br
\fBepkg [options] -c\fP
.SH VERSION
This man page documents version 2.3 of \fBepkg\fP.
.SH DESCRIPTION
\fBepkg\fP is an \fIEncap\fP package manager.  It allows the administrator
to flexibly install, remove, and manage software packages which are
commonly installed in \fI/usr/local\fP.

\fBepkg\fP supports the \fIEncap 2.1\fP package format.
.SH PACKAGE MODES
\fBepkg\fP has the following operating modes:
.TP
.B -b
Batch mode.  \fBepkg\fP will select the latest version of each package
installed in the \fIEncap\fP source directory, remove all of the older
versions, and install the latest version.
.TP
.B -c
Clean target directory.  \fBepkg\fP will recursively go through the \fIEncap\fP
target directory and remove any dangling symbolic links which point
under the \fIEncap\fP source directory.  All subdirectories of the
target directory are processed, except for the entries in the global
exclude list (and, if \fI-E\fP is specified, \fIencap.exclude\fP
entries in the target tree).
.TP
.B -k
Check mode.  For each \fIpkgspec\fP listed on the commandline, \fBepkg\fP
will check to see if the package is installed.
.TP
.B -i
Install mode.  \fBepkg\fP will install only the package(s) specified
on the commandline.  This mode is assumed if none are specified.

If versioning is on (see \fI-S\fP), \fBepkg\fP will scan the source
directory for all versions of each \fIpkgspec\fP listed on the
commandline.  It then selects one version to be installed, removes all
the other versions, and installs the selected one.  If the
\fIpkgspec\fP specifies a version, that version is selected for
installation.  If no version is specified, the latest available
version is selected.

If versioning is off, \fBepkg\fP will install only the exact package
name and version which matches each \fIpkgspec\fP on the commandline.
No other versions will be removed.
.TP
.B -r
Remove mode.  \fBepkg\fP will remove the package(s) listed on
the commandline.

If versioning is on (see \fI-S\fP), \fBepkg\fP will remove all
versions of each \fIpkgspec\fP which do not specify a version.  For
the \fIpkgspec\fPs which do specify a version, only that version is
removed.

If versioning is off, only the exact package name and version which
matches each \fIpkgspec\fP will be removed.
.TP
.B -u
Update mode.  This mode will check a set of central distribution
directories for new Encap packages and install them on the local system.
The central distribution directories are referred to as the "update path".

The update path is specified as a space-delimited list in the
\fBENCAP_UPDATE_PATH\fP environment variable.  In directory names, the
special token "%p" is replaced by the host platform name.

If one or more \fIpkgspec\fPs are listed on the commandline, the update
path is searched for newer versions of the listed packages than
those that are already available on the system.  If a listed package
is not available locally, any version of that package which is found in
the update path is installed.

If no \fIpkgspec\fPs are given on the commandline, the update path
is searched for newer versions of each of the packages which are already
available on the system.
.SH OPTIONS
Options which turn on or off features act as toggles.  Specifying them
twice will return to the default behavior.
.TP
.B -1
Back off from most recent version.  In versioning install mode, for each
\fIpkgspec\fP which doesn't specify a version, install the next-to-latest
version of the package rather than the latest version.
.TP
.B -a
Use absolute symlinks.  The default is to use relative links.
.TP
.B -A platform_suffix
Add \fIplatform_suffix\fP to the list of acceptable platform suffixes.
Packages with any of the listed suffixes are considered acceptable for
installation.
.TP
.B -C
Client operation.  \fBepkg\fP will execute package scripts, but wont do any
linking or removing.  This is useful when your Encap source and target
directories are NFS-mounted on many client machines from a central
server.  The linking or removing is taken care of on the server, but it
is often desirable to run the package scripts on each client as well.
.TP
.B -d
Tells \fBepkg\fP not to remove empty target directories.
.TP
.B -D
Tells \fBepkg\fP not to honor \fIlinkdir\fP directives in the packages
it processes.
.TP
.B -E
Check for \fIencap.exclude\fP files in places other than under the package
directory.  This includes the \fIEncap\fP source directory itself and
all directories in the \fIEncap\fP target tree.  This option is deprecated,
and is provided only for compatibility reasons.
.TP
.B -f
Force conflict resolution.  If a file already exists when
\fBepkg\fP attempts to create a link, it will be removed so that the
link can be created.
.TP
.B -F
Display debugging output for update mode.  This is especially useful for
FTP connections.
.TP
.B -H host_platform
Use \fIhost_platform\fP to identify the platform name of the machine
\fBepkg\fP is running on.  If this option is not specified, the platform
name is generated automatically.
.TP
.B -K
Keep checking all update directories to find the latest matching
package.  This option is enabled by default.  If disabled, \fBepkg\fP
will stop looking for updates after the first update directory that
contains any version of the requested package.
.TP
.B -l
Disable transaction logging.
.TP
.B -L
Tells \fBepkg\fP not to honor \fIlinkname\fP directives in the packages
it processes.
.TP
.B -n
Show what operations would have been performed without actually doing
anything.
.TP
.B -N
Removes the default entries from the global exclude list.
.TP
.B -o
Removes the default entries from the list of packages to automatically
override.

The override list is primarily intended to help people transition from
a monolithic \fI/usr/local\fP to using \fIEncap\fP.
.TP
.B -O pkgspec
Add \fIpkgspec\fP to the list of packages whose links will be overriden
when a new package is installed.  This option can be used multiple times.
.TP
.B -p
Don't perform prerequisite checking.
.TP
.B -P
Link or remove files directly in the package directory.  Normally, files
in the package directory are not linked or removed.  Special package
files, such as \fIencapinfo\fP and the package scripts, are not
linked or removed even when \fB-P\fP is specified.  This option is deprecated,
and is provided only for compatibility reasons.
.TP
.B -q
Quiet mode.  \fBepkg\fP will not produce any output except for error
messages.
.TP
.B -R
Don't run package scripts.
.TP
.B -s source_dir
Tells \fBepkg\fP to use \fIsource_dir\fP as the \fIEncap\fP source
directory.  If the source is not specified but the target is,
\fBepkg\fP will assume the relative path from the target to the
source is \fIencap/\fP.  If neither is specified, \fBepkg\fP will check the
contents of the \fPENCAP_SOURCE\fP environment variable.  If that is
not set, it will use the default set at compile-time (usually
\fI/usr/local/encap\fP).
.TP
.B -S
Specific package only.  This tells \fIepkg\fP to disable versioning
and operate only on the exact package and version specified.

This option is only meaningful in install and remove modes.
.TP
.B -t target_dir
Tells \fBepkg\fP to use \fItarget_dir\fP as the \fIEncap\fP target
directory.  If the target is not specified but the source is,
\fBepkg\fP will assume the relative path from the source to the
target is \fI../\fP.  If neither is specified, \fBepkg\fP will check the
contents of the \fBENCAP_TARGET\fP environment variable.  If that is
not set, it will use the default set at compile-time (usually
\fI/usr/local\fP).
.TP
.B -T
Don't automatically retry connections to update directories.  The
default is to retry connections.
.TP
.B -U update_dir
Prepends \fIupdate_dir\fP to the update path (which is specified by the
\fIENCAP_UPDATE_PATH\fP environment variable).
.TP
.B -v
Increase verbosity.  This option can be specified multiple times.
.TP
.B -V
Print \fBepkg\fP's version and compile-time defaults.
.TP
.B -x
Do not check for excluded files within packages.  This does not
affect the global exclude list (see \fI-X\fP) or \fIencap.exclude\fP files
in other places (see \fI-E\fP).
.TP
.B -X exclude_path
Add \fIexclude_path\fP to the global exclude list.

If the \fIEncap\fP source directory lies under the \fIEncap\fP target
directory, the relative path from target to source is automaticly added
to the list.
.SH TRANSACTION LOGGING
By default, \fBepkg\fP will log an entry to the file \fIepkg.log\fP in
the \fIEncap\fP source directory for each transaction indicating the date,
target directory, package, install or remove as appropriate, and the
status.  The status is \fIsuccess\fP if completely successful,
\fIfailed\fP if there was a fatal error, \fIforced\fP if invoked in
force mode (see above), or \fIpartial\fP if the operation was
partially successful (for example, most of the links were created, but
some could not be created because of a link by the same name into a
previously-installed package).
.SH EXAMPLES
To install only package \fImutt-0.93.2i\fP:

	\fBepkg -Si mutt-0.93.2i\fP

To upgrade to the latest version of \fImutt\fP:

	\fBepkg mutt\fP

.SH BUGS
Probably quite a few.  Reports are welcome.
.SH ENVIRONMENT VARIABLES
.B ENCAP_SOURCE
.br
.B ENCAP_TARGET
.br
.B ENCAP_PKGNAME
.br
.B ENCAP_MODE
.br
.B ENCAP_UPDATE_PATH
.SH FILES
.I /usr/local/encap/pkgname/encapinfo
.br
.I /usr/local/encap/pkgname/preinstall
.br
.I /usr/local/encap/pkgname/postinstall
.br
.I /usr/local/encap/pkgname/preremove
.br
.I /usr/local/encap/pkgname/postremove
.SH SEE ALSO
.BR mkencap (1)
.br
The \fBepkg\fP Homepage (\fIhttp://www.encap.org/epkg/\fP)
.br
The Encap Archive (\fIhttp://www.encap.org/\fP)
.SH AUTHOR
Mark D. Roth <roth@feep.net>


NEWS
User-visible changes from 2.2.x to 2.3.x:
-----------------------------------------

* added HTTP download support using cURL (see README file for details)

* added automatic reconnection if FTP server times out between
  downloads; use "-T" option to disable

* new package profile format used by mkencap for automated package creation
  (see encap_profile(5) man page for explanation of the profile format)
  (requires libexpat to be installed at build time and GNU m4 to be
  installed at run-time; see README for details)

* removed mkencap directory creation mode ("mkencap -i")

* many changes to mkencap command line interface

* added encapinfo_parse_directive() to libencap API

* changed Encap 2.1 platform name for HP-UX from "pa1.1-hpux11.00"
  to "pa-hpux11.00"
  (encap_platform_compat() will still recognize the old name)



epkg/epkg.c
/*
**  Copyright 1998-2003 University of Illinois Board of Trustees
**  Copyright 1998-2004 Mark D. Roth
**  All rights reserved.
**
**  epkg.c - main driver program for epkg
**
**  Mark D. Roth <roth@feep.net>
*/

#include <epkg.h>

#include <stdio.h>
#include <sys/param.h>
#include <errno.h>

#ifdef STDC_HEADERS
# include <string.h>
# include <stdlib.h>
#endif

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#ifdef HAVE_LIBTAR
# include <libtar.h>
#endif

#ifdef HAVE_LIBZ
# include <zlib.h>
#endif

#ifdef HAVE_LIBFGET
# include <libfget.h>
#endif

#ifdef HAVE_LIBCURL
# include <curl/curl.h>
#endif


/* global variables */
char platform[MAXPATHLEN];
unsigned short verbose = 1;
unsigned long options = OPT_DEFAULTS;
unsigned long epkg_opts = (EPKG_OPT_WRITELOG | EPKG_OPT_VERSIONING | EPKG_OPT_UPDATEALLDIRS);
encap_list_t *exclude_l = NULL;
encap_list_t *override_l = NULL;
encap_list_t *platform_suffix_l = NULL;


static void
usage(void)
{
	fprintf(stderr, "Options: [-1aCdDEfFlLnNopPqRSTvVx] "
		"[-s source] [-t target]\n");
	fprintf(stderr, "         [-O overrides] [-X excludes] "
		"[-U update_dir]\n");
	fprintf(stderr, "         [-H host_platform] [-A platform_suffix]\n");
	fprintf(stderr, "Usage:\n");
	fprintf(stderr, "   epkg [options] [-i] pkgspec ...\n");
	fprintf(stderr, "   epkg [options] -r pkgspec ...\n");
	fprintf(stderr, "   epkg [options] -k pkgspec ...\n");
	fprintf(stderr, "   epkg [options] -u [pkgspec ...]\n");
	fprintf(stderr, "   epkg [options] -b\n");
	fprintf(stderr, "   epkg [options] -c\n");
	exit(1);
}


static void
print_defaults(void)
{
	printf("epkg %s by Mark D. Roth <roth@feep.net>\n", PACKAGE_VERSION);
	printf("   Build Platform:\t\t%s\n", PLATFORM);
#ifdef HAVE_LIBTAR
	printf("   Builtin tar:\t\t\tlibtar-%s\n", libtar_version);
#else
	printf("   Builtin tar:\t\t\tN/A\n");
#endif
#ifdef HAVE_LIBZ
	printf("   Builtin gzip:\t\tzlib-%s\n", ZLIB_VERSION);
#else
	printf("   Builtin gzip:\t\tN/A\n");
#endif
#ifdef HAVE_LIBFGET
	printf("   Builtin FTP:\t\t\tlibfget-%s\n", libfget_version);
#else
	printf("   Builtin FTP:\t\t\tN/A\n");
#endif
#ifdef HAVE_LIBCURL
	printf("   Builtin HTTP:\t\tlibcurl-%s\n", LIBCURL_VERSION);
#else
	printf("   Builtin HTTP:\t\tN/A\n");
#endif
	printf("   Default Encap Source:\t%s\n", DEFAULT_SOURCE);
	printf("   Default Encap Target:\t%s\n", DEFAULT_TARGET);
	printf("   Default Exclude List:\t%s\n", DEFAULT_EXCLUDES);
	printf("   Default Override List:\t%s\n", DEFAULT_OVERRIDES);
	exit(0);
}


/* main() */
int
main(int argc, char *argv[])
{
	char buf[MAXPATHLEN], cwd[MAXPATHLEN];
	encap_listptr_t lp;
	short pkg_mode = 0;
	int i, j, exit_val = 0;
	char *opt_src = NULL, *opt_tgt = NULL;

	/* initialize exclude list */
	exclude_l = encap_list_new(LIST_USERFUNC, NULL);
	encap_list_add_str(exclude_l, DEFAULT_EXCLUDES, ":");

	/* initialize override list */
	override_l = encap_list_new(LIST_USERFUNC, NULL);
	encap_list_add_str(override_l, DEFAULT_OVERRIDES, ":");

	/* initialize update directory list */
	if (update_path_init() == -1)
	{
		perror("malloc()");
		exit(1);
	}

	/* initialize compatible platform suffix list */
	encap_platform_name(platform, sizeof(platform));
	platform_suffix_l = encap_list_new(LIST_QUEUE, NULL);

	/* parse arguments */
	while ((i = getopt(argc, argv,
			   "1aA:bcCdDEfFhH:ikKlL"
			   "nNoO:pPqrRs:St:TuU:vVxX:")) != -1)
	{
		switch (i)
		{
		case 'V':
			print_defaults();
		case 'u':
			if (pkg_mode)
				usage();
			pkg_mode = MODE_UPDATE;
			BIT_SET(epkg_opts, EPKG_OPT_VERSIONING);
			BIT_CLEAR(epkg_opts, EPKG_OPT_BACKOFF);
			break;
		case 'b':
			if (pkg_mode)
				usage();
			pkg_mode = MODE_BATCH;
			break;
		case 'i':
			if (pkg_mode)
				usage();
			pkg_mode = MODE_INSTALL;
			break;
		case 'r':
			if (pkg_mode)
				usage();
			pkg_mode = MODE_REMOVE;
			break;
		case 'k':
			if (pkg_mode)
				usage();
			pkg_mode = MODE_CHECK;
			break;
		case 'c':
			if (pkg_mode)
				usage();
			pkg_mode = MODE_CLEAN;
			break;
		case 'q':
			verbose = 0;
			break;
		case 'v':
			verbose++;
			break;
		case 'S':
			BIT_TOGGLE(epkg_opts, EPKG_OPT_VERSIONING);
			break;
		case '1':
			BIT_TOGGLE(epkg_opts, EPKG_OPT_BACKOFF);
			break;
		case 'l':
			BIT_TOGGLE(epkg_opts, EPKG_OPT_WRITELOG);
			break;
		case 'K':
			BIT_TOGGLE(epkg_opts, EPKG_OPT_UPDATEALLDIRS);
			break;
		case 'F':
			BIT_TOGGLE(dl_opts, DL_OPT_DEBUG);
			break;
		case 'T':
			BIT_TOGGLE(dl_opts, DL_OPT_CONNRETRY);
			break;
		case 'E':
			BIT_TOGGLE(epkg_opts, EPKG_OPT_OLDEXCLUDES);
			BIT_TOGGLE(options, OPT_TARGETEXCLUDES);
			break;
		case 'L':
			BIT_TOGGLE(options, OPT_LINKNAMES);
			break;
		case 'd':
			BIT_TOGGLE(options, OPT_NUKETARGETDIRS);
			break;
		case 'D':
			BIT_TOGGLE(options, OPT_LINKDIRS);
			break;
		case 'f':
			BIT_TOGGLE(options, OPT_FORCE);
			break;
		case 'n':
			BIT_TOGGLE(options, OPT_SHOWONLY);
			break;
		case 'a':
			BIT_TOGGLE(options, OPT_ABSLINKS);
			break;
		case 'R':
			BIT_TOGGLE(options, OPT_RUNSCRIPTS);
			break;
		case 'C':
			BIT_TOGGLE(options, OPT_RUNSCRIPTSONLY);
			break;
		case 'x':
			BIT_TOGGLE(options, OPT_EXCLUDES);
			break;
		case 'p':
			BIT_TOGGLE(options, OPT_PREREQS);
			break;
		case 'P':
			BIT_TOGGLE(options, OPT_PKGDIRLINKS);
			break;
		case 'N':
			encap_list_free(exclude_l, free);
			exclude_l = encap_list_new(LIST_USERFUNC, NULL);
			break;
		case 'H':
			strlcpy(platform, optarg, sizeof(platform));
			break;
		case 'A':
			encap_list_add(platform_suffix_l, strdup(optarg));
			break;
		case 'X':
			encap_list_add(exclude_l, strdup(optarg));
			break;
		case 'o':
			encap_list_free(override_l, free);
			override_l = encap_list_new(LIST_USERFUNC, NULL);
			break;
		case 'O':
			encap_list_add(override_l, strdup(optarg));
			break;
		case 'U':
			update_path_add(optarg, 1);
			break;
		case 's':
			opt_src = strdup(optarg);
			break;
		case 't':
			opt_tgt = strdup(optarg);
			break;
		case 'h':
		default:
			usage();
		}
	}

	/* set verbosity for download and archive modules */
	dl_verbose = verbose;
	archive_verbose = verbose;

	/* set default mode if not specified */
	if (pkg_mode == 0)
		pkg_mode = MODE_INSTALL;

	/* make sure we have the right number of arguments for this mode */
	i = argc - optind;
	switch (pkg_mode)
	{
	case MODE_UPDATE:
		/* arguments are optional in update mode */
		break;
	case MODE_CHECK:
	case MODE_REMOVE:
	case MODE_INSTALL:
		/* need arguments for install, remove, and check modes */
		if (i < 1)
			usage();
		break;
	case MODE_CLEAN:
	case MODE_BATCH:
		/* don't accept arguments for batch or clean modes */
		if (i >= 1)
			usage();
		break;
	}

	if (getcwd(cwd, sizeof(cwd)) == NULL)
	{
		fprintf(stderr, "epkg: getcwd(): %s\n", strerror(errno));
		exit(1);
	}
	init_encap_paths(cwd, opt_src, opt_tgt);

	/* if source dir is under target dir, add it to the exclude list */
	encap_relativepath(target, source, buf, sizeof(buf));
	if (strncmp(buf, "..", 2) != 0)
		encap_list_add(exclude_l, strdup(buf));

	/* print out our options */
	if (verbose > 3)
	{
		printf("epkg: printing options:\n");
		printf("  * source directory:\t\t%s\n", source);
		printf("  * target directory:\t\t%s\n", target);
		printf("  * update path:\t\t");
		update_path_print();
		putchar('\n');
		printf("  * global exclude list:\t");
		encap_listptr_reset(&lp);
		while (encap_list_next(exclude_l, &lp) != 0)
			printf("%s%s", (exclude_l->first == lp ? "" : ":"),
			       (char *)encap_listptr_data(&lp));
		if (exclude_l->first == NULL)
			printf("N/A");
		putchar('\n');
		printf("  * override list:\t\t");
		encap_listptr_reset(&lp);
		while (encap_list_next(override_l, &lp) != 0)
			printf("%s%s", (override_l->first == lp ? "" : ":"),
			       (char *)encap_listptr_data(&lp));
		if (override_l->first == NULL)
			printf("N/A");
		putchar('\n');
		printf("  * host platform:\t\t%s\n", platform);
		printf("  * platform suffix list:\t");
		encap_listptr_reset(&lp);
		while (encap_list_next(platform_suffix_l, &lp) != 0)
			printf("%s%s",
			       (platform_suffix_l->first == lp ? "" : ":"),
			       (char *)encap_listptr_data(&lp));
		if (platform_suffix_l->first == NULL)
			printf("N/A");
		putchar('\n');
		printf("  * perform logging:\t\t%s\n",
		       (BIT_ISSET(epkg_opts, EPKG_OPT_WRITELOG) ? "yes" :
			"no"));
		printf("  * handle multiple versions:\t%s\n",
		       (BIT_ISSET(epkg_opts, EPKG_OPT_VERSIONING) ? "yes" :
			"no"));
		printf("  * default version:\t\t%slatest\n",
		       (BIT_ISSET(epkg_opts, EPKG_OPT_BACKOFF) ? "second-" :
			""));
		printf("  * update debug output:\t%s\n",
		       (BIT_ISSET(dl_opts, DL_OPT_DEBUG) ? "yes" :
			"no"));
		printf("  * read legacy excludes:\t%s\n",
		       (BIT_ISSET(epkg_opts, EPKG_OPT_OLDEXCLUDES) ? "yes" :
			"no"));
		printf("  * link style:\t\t\t%s\n",
		       (BIT_ISSET(options, OPT_ABSLINKS) ? "absolute" :
			"relative"));
		printf("  * check package prereqs:\t%s\n",
		       (BIT_ISSET(options, OPT_PREREQS) ? "yes" : "no"));
		printf("  * run package scripts:\t%s\n",
		       (BIT_ISSET(options, OPT_RUNSCRIPTS) ? "yes" : "no"));
		printf("  * honor package exclusions:\t%s\n",
		       (BIT_ISSET(options, OPT_EXCLUDES) ? "yes" : "no"));
		printf("  * link files in pkg dir:\t%s\n",
		       (BIT_ISSET(options, OPT_PKGDIRLINKS) ? "yes" : "no"));
		printf("  * remove empty target dirs:\t%s\n",
		       (BIT_ISSET(options, OPT_NUKETARGETDIRS) ? "yes" :
			"no"));
		printf("  * force mode:\t\t\t%s\n",
		       (BIT_ISSET(options, OPT_FORCE) ? "yes" : "no"));
		printf("  * display actions only:\t%s\n",
		       (BIT_ISSET(options, OPT_SHOWONLY) ? "yes" : "no"));
		printf("  * process package links:\t%s\n",
		       (BIT_ISSET(options, OPT_RUNSCRIPTSONLY) ? "no" :
			"yes"));
		printf("  * package mode:\t\t");
		switch (pkg_mode)
		{
		case MODE_UPDATE:
			printf("update");
			break;
		case MODE_INSTALL:
			printf("install");
			break;
		case MODE_REMOVE:
			printf("remove");
			break;
		case MODE_BATCH:
			printf("batch");
			break;
		case MODE_CHECK:
			printf("check");
			break;
		case MODE_CLEAN:
			printf("clean");
			break;
		}
		printf("\n\n");
	}

	switch (pkg_mode)
	{

	case MODE_CLEAN:
		exit_val = clean_mode();
		break;

	case MODE_UPDATE:
		if (!update_path_isset())
		{
			fprintf(stderr, "epkg: no update path set\n");
			exit(1);
		}
		if (putenv("ENCAP_MODE=update") != 0)
		{
			perror("putenv()");
			exit(1);
		}
		if (argc - optind == 0)
		{
			exit_val = update_mode(NULL);
			break;
		}
		for (i = 0; i < argc - optind; i++)
		{
			/* download and/or untar the new package */
			j = update_mode(argv[optind + i]);
			if (j)
			{
				exit_val++;
				if (j == -1)
					break;
			}
		}
		break;

	case MODE_CHECK:
		for (i = 0; i < argc - optind; i++)
		{
			strcpy(buf, basename(argv[optind + i]));
			if (check_mode(buf) != 0)
				exit_val++;
			if (verbose > 1)
				putchar('\n');
		}
		break;

	case MODE_INSTALL:
		if (putenv("ENCAP_MODE=install") != 0)
		{
			perror("putenv()");
			exit(1);
		}
		for (i = 0; i < argc - optind; i++)
		{
			j = install_mode(argv[optind + i]);
			if (j != 0)
			{
				exit_val++;
				if (j == -1)
					break;
			}
			if (verbose > 1)
				putchar('\n');
		}
		break;

	case MODE_REMOVE:
		if (putenv("ENCAP_MODE=remove") != 0)
		{
			perror("putenv()");
			exit(1);
		}
		for (i = 0; i < argc - optind; i++)
		{
			strcpy(buf, basename(argv[optind + i]));
			if (remove_mode(buf) != 0)
				exit_val++;
			if (verbose > 1)
				putchar('\n');
		}
		break;

	case MODE_BATCH:
		if (putenv("ENCAP_MODE=batch") != 0)
		{
			perror("putenv()");
			exit(1);
		}
		exit_val = batch_mode();
		if (verbose > 1)
			putchar('\n');
		break;

	} /* switch() */

	/* clean up */
	download_cleanup();
	update_path_free();

	exit(exit_val);
}


